### PR TITLE
Add on_secondary_use for items, only triggers if right clicking on sky

### PR DESCRIPTION
**Pull request by thePalindrome**
Note that this reserves action 5 to do so. This has been tested with a quick lua script.

### CODE REVIEW

File: builtin/game/item.lua [hunk 1]
```diff
@@ -605,6 +609,7 @@ core.craftitemdef_default = {
 	-- Interaction callbacks
 	on_place = redef_wrapper(core, 'item_place'), -- core.item_place
 	on_drop = redef_wrapper(core, 'item_drop'), -- core.item_drop
+	on_rightclick = redef_wrapper(core, 'item_rightclick'),
```
Comment (est31) [inline builtin/game/item.lua]: perhaps rather call it `on_right_click`?

File: src/script/cpp_api/s_item.cpp [hunk 2]
```diff
@@ -110,6 +110,32 @@ bool ScriptApiItem::item_OnUse(ItemStack &item,
...
+	if (!getItemCallback(item.name.c_str(), "on_rightclick"))
+		return false;
+	
+	LuaItemStack::create(L, item);
+	objectrefGetOrCreate(L, user);
+	PointedThing pointed;
+	pointed.type = POINTEDTHING_NOTHING;
+	pushPointedThing(pointed);
+	PCALL_RES(lua_pcall(L, 3, 1, error_handler));
+	if(!lua_isnil(L, -1)) {
```
Comment (est31) [inline src/script/cpp_api/s_item.cpp]: space

File: src/game.cpp [hunk 3]
```diff
@@ -3618,6 +3621,15 @@ void Game::processPlayerInteraction(std::vector<aabb3f> &highlight_boxes,
 }
 
 
+void Game::handlePointingAtNothing(GameRunData *runData, const ItemStack &playerItem)
+{
+	infostream << "Right Clicked in Air" << std::endl;
+	PointedThing fauxPointed;
+	fauxPointed.type = POINTEDTHING_NOTHING;
+	client->interact(5,fauxPointed);
```
Comment (est31) [inline src/game.cpp]: space

File: builtin/game/item.lua (hunk 1 above)
Comment (thePalindrome) [inline builtin/game/item.lua]: The function is titled on_rightclick in entities, otherwise I would've gone with on_right_click

**Comment by est31** [comment]
Looks already very good for a first contribution.

It successfully gets backwards compatibility with older network versions: old clients won't send the action to new servers, so there is no problem there, and new clients will send the action to old servers, but the server silently ignores actions it doesn't know, with an entry into infostream, which is acceptable. So okay from this side.

I've found some minor style issues about whitespaces here and there, I've noted them above.

It should be `right_click` inside the lua API, as here we can't change anything, so its better to get it separated now, before one regrets it it later on.

About a general design notice, I think it would be best if on_rightclick got merged with on_place. It would be much less intrusive, only the client and protocol parts and a little server part will remain. Also, the calls are already very similar, and we even already have a type for pointed_thing that's none. So it would go easily.

File: builtin/game/item.lua (hunk 1 above)
Comment (est31) [inline builtin/game/item.lua]: Okay then, that's a point.

**Comment by est31** [comment]
Looks good :+1:

**Comment by est31** [comment]
Looks good :+1:

**Comment by paramat** [comment]
Maybe it would help this PR if you can find others who also need this feature. Can you describe situations where this is useful?

**Comment by rubenwardy** [comment]
I could see this being useful for a gun mod, to active sniper mode.

**Comment by thePalindrome** [comment]
One of the modders in the #minetest channel mentioned that he had a hook to check for a right-click every frame to detect right click for firing a weapon at a distant target. Other such uses would be an item that would be expected not to be pointed at a node (think some flight item).

**Comment by paramat** [comment]
Okay thanks sounds useful i support the feature, not an expert on the code though.

**Comment by CasimirKaPazi** [comment]
I have several uses for this. One is for the subgame UFOrun. I wanted that in creative you would be able to place nodes in the air. (Maybe where you stand, maybe in fixed distance.) With current minetest you would have to use on_use(), but blocking the ability to dig with this. With this pull request, I can just overwrite the default node definition to do special placing on_secondary_use().

It is even better then wished back then, because it easily allows different uses for different cases. E.g. place where you stand on_secondary_use, and place normal on_place.

**Comment by Ferk** [comment]
This would be useful for the magic scrolls in Dungeontest as well.

**Comment by paramat** [comment]
We just need someone to confirm the code is good and give another +1.

**Comment by paramat** [comment]
I'll merge this later it has 2 approvals and it seems est31 is happy with the code.

**Comment by paramat** [comment]
97908cc65670d3f6cf2e286390bfea10f653aaa8

**Pull request closed without merging**